src/kaleidoscope/MatrixAddr.h

```diff
@@ -0,0 +1,385 @@
+    this->row_ = other.row();
+    this->col_ = other.col();
+    return *this;
+  }
+
+  template<typename MatrixAddr__>
+  MatrixAddr &operator+(const MatrixAddr__ &other) {
+    this->row_ += other.row();
+    this->col_ += other.col();
+    return *this;
+  }
+
+  template<typename MatrixAddr__>
+  MatrixAddr &operator-(const MatrixAddr__ &other) {
+    this->row_ -= other.row();
+    this->col_ -= other.col();
+    return *this;
+  }
+
+  template<typename MatrixAddr__>
+  MatrixAddr &operator+=(const MatrixAddr__ &other) {
+    *this = *this + other;
+    return *this;
+  }
+
+  template<typename MatrixAddr__>
+  MatrixAddr &operator-=(const MatrixAddr__ &other) {
+    *this = *this - other;
+    return *this;
+  }
+};
+
+// This is a copressed storage version of MatrixAddr.
+// It uses a common integer to store row and column indices.
+// To enable fast row/col access, bitshifts are used instead of
+// integer divisions to compute row indices. This is only possible
+// by padding rows with unused entries to gain a row size that
+// is a power of two.
+//
+template<uint8_t rows__, uint8_t cols__, typename StorageType__ = uint8_t>
+class MatrixAddrCompressed {
+ private:
+
+  StorageType__ raw_;
+
+ public:
+
+  typedef MatrixAddrCompressed<rows__, cols__, StorageType__> ThisType;
+
+  // Row size must be a power of two to enable fast index extraction.
+  static constexpr uint8_t rowSize
+    = internal::matrix_addr::upperPowerOfTwo(cols__);
+
+  static constexpr StorageType__ maxRawValue
+    = internal::matrix_addr::IntLimitsTrait<StorageType__>::max;
+  static constexpr StorageType__ invalidState = maxRawValue;
+  static constexpr uint8_t shift = internal::matrix_addr::log2(rowSize);
+
+  static constexpr uint8_t rows = rows__;
+  static constexpr uint8_t cols = cols__;
+
+  static constexpr uint8_t matrixSize = rows__ * cols__;
+
+  static constexpr StorageType__ rowMask = StorageType__(maxRawValue << shift);
+  static constexpr StorageType__ colMask = StorageType__(~rowMask);
+
+  static_assert(rows * rowSize < maxRawValue, "Please use a StorageType__ with a higher capacity");
+
+  constexpr MatrixAddrCompressed() : raw_(invalidState) {}
+
+  explicit constexpr MatrixAddrCompressed(uint8_t row, uint8_t col) : raw_(row * rowSize + col) {}
+
+  explicit constexpr MatrixAddrCompressed(uint8_t offset)
+    : MatrixAddrCompressed(offset / cols, offset % cols) {}
+
+  constexpr MatrixAddrCompressed(const ThisType &other) : raw_(other.raw_) {}
+
+  template<typename MatrixAddr__>
+  constexpr MatrixAddrCompressed(const MatrixAddr__ &other)
+    : MatrixAddrCompressed(other.row(), other.col()) {}
+
+  constexpr uint8_t row() const {
+    return raw_ >> shift;
+  }
+  constexpr uint8_t col() const {
+    return raw_ & colMask;
+  }
+
+  void setRow(uint8_t r) {
+    raw_ = (r << shift) | (raw_ & colMask);
+  }
+  void setCol(uint8_t c) {
+    raw_ = (raw_ & rowMask) | c;
+  }
+
+  constexpr uint8_t offset() {
```

Comment:
I find `offset()` to be unintuitive, and this is the function I would need when using a bitfield with one bit per key on the keyboard. I think `index()` would be a much better name.

Reply:
I both agree and disagree. Index is not unique as the term is also used as row- or column-index.
In the context of the firmware the term seems to have a special meaning. There are quite a lot of functions and methods that accept a one-based parameter called index. There the value zero has a special meaning.

Reply:
The term `offset` implies to me that it is a constant, not something that would vary from one `KeyAddr` to another.

Reply:
I am not insisting on the naming `offset`. Is there maybe something else apart from `index` that we might name it?